hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java

```diff
@@ -1163,14 +1105,60 @@ void notifyNamenodeDeletedBlockInt(
       // In case of un/successful block reports we have to inform the leader that
       // block reporting has finished for now.
       if (blkReportHander != null) {
-        getLeaderActor().blockReportCompleted(bpRegistration);
+        BPServiceActor leader =  getLeaderActor();
+        if( leader != null){
+          leader.blockReportCompleted(bpRegistration);
+        }
       }
     }
 
     scheduleNextBlockReport(startTime);
     return cmds.size() == 0 ? null : cmds;
   }
 
+  StorageBlockReport[] slimBlockReports(StorageBlockReport[] reports){
+    StorageBlockReport[] slimStorageReports = new StorageBlockReport[reports.length];
+
+    for(int i = 0 ; i < reports.length; i++){
+      StorageBlockReport fatSR = reports[i];
+      BlockReport fatBR = fatSR.getReport();
+      Bucket slimBuckets[] = new Bucket[fatBR.getBuckets().length];
+
+      for(int j = 0; j < fatBR.getBuckets().length; j++){
+        Bucket slimBucket = new Bucket();
+        slimBucket.setHash(fatBR.getBuckets()[j].getHash());
+        slimBucket.setBlocks(BlockListAsLongs.EMPTY);
+        slimBuckets[j] = slimBucket;
+      }
+
+      BlockReport slimBR = new BlockReport(slimBuckets, fatBR.getNumberOfBlocks());
+
+      slimStorageReports[i] = new StorageBlockReport(fatSR.getStorage(), slimBR);
+    }
+
+    return slimStorageReports;
+  }
+
+  public void removeMatchingBuckets(DatanodeCommand cmd, StorageBlockReport reports[]){
+    HashesMismatchCommand hmcmd = (HashesMismatchCommand) cmd;
+    Map<String, List<Integer>> map = ((HashesMismatchCommand) cmd).getMissMatchingBuckets();
+    for(StorageBlockReport report : reports){
+      List<Integer> mismatchingBuckets = map.get(report.getStorage().getStorageID());
+      if(mismatchingBuckets != null && mismatchingBuckets.size()>0){
```

Comment:
won't this make that if all the bucket are matching it sends a full block report?

Reply:
no

Reply:
You will need to explain more than that.
The way I read it the code does the following:
get a report for a storage, that I assume contains all the bucket.
get the mismatching bucket for the storage associated to this report.
if the number of mismatching buckets is more than 0:
   go through all the buckets in the report, mark to skip and set an empty block list for all the buckets that are not mismatching.

So if there are no mismatching buckets none of the matching buckets will be marked as skip and it will send all the blocks (full block report).
What am I missing?